metrics/store.go

```diff
@@ -366,12 +367,41 @@ func (s *Store) metricKey(envelope *events.Envelope) string {
 		buffer.WriteString(envelope.GetContainerMetric().GetApplicationId())
 		buffer.WriteString(strconv.Itoa(int(envelope.GetContainerMetric().GetInstanceIndex())))
 	case events.Envelope_CounterEvent:
-		buffer.WriteString(envelope.GetCounterEvent().GetName())
+		buffer.WriteString(generateMetricName(envelope, "CounterEvent"))
 	case events.Envelope_HttpStartStop:
 		buffer.WriteString(envelope.GetHttpStartStop().GetRequestId().String())
 	case events.Envelope_ValueMetric:
-		buffer.WriteString(envelope.GetValueMetric().GetName())
+		buffer.WriteString(generateMetricName(envelope, "ValueMetric"))
 	}
 
 	return buffer.String()
 }
+
+// Issue:
+// 		https://github.com/bosh-prometheus/firehose_exporter/issues/47
+// Solution:
+// 1. Check if the metric being emitted comes from the container of the app
+// the way we are going to do that is by checking if the origin has a uuid
+// which means its from the container
+// 2. Once detected that is from the app container, then add a little more extra info
+// on the metric name like tags, just to differentiate the metrics and emit all of them
+func generateMetricName(envelope *events.Envelope, whichType string) string {
+	var name string
+	if whichType == "ValueMetric" {
+		name = envelope.GetValueMetric().GetName()
+	} else {
+		name = envelope.GetCounterEvent().GetName()
+	}
+	uuidRegex := "^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[8|9|aA|bB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}$"
+	reg := regexp.MustCompile(uuidRegex)
+	isItValidUuid := reg.MatchString(*envelope.Origin)
+	if isItValidUuid {
+		if envelope.Tags != nil {
+			for _, tagValue := range envelope.Tags {
```

Comment:
This change is almost exactly what we were looking for to resolve the issue of not getting all custom metrics. However by looping over a map the order is nondeterministic and results in duplicate metrics ending up in the metricStore - thereby resulting on scrape time errors. This can be resolved by first sorting by the tag names and ensuring a deterministic order for generation of the key. 

This has been implemented here along with a WIP implementation of allowing operators to selectively change the Origin of application metrics for cleaner consumption in prometheus/grafana upstream - https://github.com/dickssportinggoods/firehose_exporter/commit/64fd5f2e92681057a2e27916bc7b4834f51548c7

```
		if envelope.Tags != nil {
			// These tags need to be sorted before iterating over them as iterating
			// over a map results in nondeterministic keys being used as a metricKey.
			sortedTags := make([]string, 0, len(envelope.Tags))
			for tag := range envelope.Tags {
				sortedTags = append(sortedTags, tag)
			}
			sort.Strings(sortedTags)
			for _, tag := range sortedTags {
				name = name + string([]byte{255}) + utils.NormalizeName(envelope.Tags[tag])
			}
		}
```